### PR TITLE
feat: estimator for AMM prices

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
## Description

An estimator for the AMM Dapp to use to estimate prices, including fees and slippage.

It's completely untested.

### Security Considerations

Not a security issue.

### Documentation Considerations

This is the simplest description I've written to date on how the AMM calculates prices.

### Testing Considerations

This should be tested.

### CODE REVIEW

File: packages/run-protocol/src/vpool-xyk-amm/constantProduct/estimate.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  const estimateProceeds = (amountIn, brandOut, pools) => {
+    let proceeds;
+    if (amountIn.brand === centralBrand || brandOut === centralBrand) {
+      proceeds = estimateSinglePoolProceeds(amountIn, brandOut, pools);
+    } else {
+      proceeds = estimateDoublePoolProceeds(amountIn, brandOut, pools);
+    }
+    return reduceBy(proceeds, slippageBP);
```
Comment (samsiegart) [inline packages/run-protocol/src/vpool-xyk-amm/constantProduct/estimate.js]: This should be `reduceBy(slippageBP, proceeds)`

File: packages/run-protocol/src/vpool-xyk-amm/constantProduct/estimate.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    const outRequired = estimateSinglePoolRequired(
+      centralBrand,
+      amountOut,
+      pools,
+    );
+    const inRequired = estimateSinglePoolRequired(brandIn, outRequired, pools);
+    return inRequired;
+  }
+
+  const estimateProceeds = (amountIn, brandOut, pools) => {
```
Comment (samsiegart) [inline packages/run-protocol/src/vpool-xyk-amm/constantProduct/estimate.js]: This doesn't seem to be returning the correct values. The return value is orders of magnitude larger than I'd expect, and seems to be returning an amount with the "amountIn" brand, not the "brandOut" brand,

Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/constantProduct/estimate.js]: As Dan said recently, "you don't have to test all of your code, just the part you want to rely on to be correct."

File: packages/run-protocol/src/vpool-xyk-amm/estimate.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+const makeEstimator = (centralBrand, rates) => {
+  const { poolFeeBP, protocolFeeBP, slippageBP } = rates;
+
+  // Proceeds less slippage when specifying the amount to add to a single pool
+  function estimateSinglePoolProceeds(amountIn, brandOut, pools) {
+    const isCentralIn = amountIn.brand === centralBrand;
+    const pool = isCentralIn ? pools.get(brandOut) : pools.get(amountIn.brand);
+    const kPre =
+      pool.getSecondaryAmount().value * pool.getCentralAmount().value;
```
Comment (samsiegart) [inline packages/run-protocol/src/vpool-xyk-amm/estimate.js]: Could this be `pool.secondaryAmount.value` instead (and similarly for `pool.centralAmount.value`)? That way the interface would match https://github.com/Agoric/agoric-sdk/blob/434a2408a9455b1ace2956eb541b7b99cb6fbc17/packages/run-protocol/src/vpool-xyk-amm/pool.js#L60-L63 which the UI has reference to.

Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/estimate.js]: Done.

**Review by samsiegart — APPROVED** [review]
LGTM, seems to be working great on https://github.com/Agoric/dapp-amm/tree/pool-subscriptions

**Pull request merged**